rfcs/text/refactoring-date-component.md

```diff
@@ -0,0 +1,398 @@
+
+As the number of consumers of the library has grown this has led to a lot of [issues](https://github.com/Sage/carbon/issues?q=is%3Aissue+is%3Aopen+date) being raised against the component. Some of the code was written over four years ago, and the addition of some newer features during this time have led to the code for the component being significantly bloated and therefore difficult to maintain.
+
+# Detailed design
+
+The date passed to the picker component can be updated through a `useEffect` hook that checks the current value in the input is both a valid format (based on the value recieved from the locale context) and parses to a valid Date. As such the component will be able to maintain a significantly reduced amount of internal state for the `selectedDate` and the `open` state for the picker component. The `DatePicker` will be baked into the `Date` component and we will surface the interface defined in the [react-day-picker](https://react-day-picker.js.org/api/DayPicker) docs via a `pickerProps` object and as such it can be moved to an `__internal__` directory: we will also include a link to `react-day-picker` documentations in our prop tables.
+
+```jsx
+  // src/components/date/date.component.js
+  import {
+    formatToString,
+    isValidFormat,
+    parseToDate,
+    isValidDate,
+    formatISOString,
+    localeMap,
+  } from "./__internal__/date-utils"
+
+  const DateInput = ({ value, onChange, onBlur, onFocus, inputProps, pickerProps, ...rest }) => {
+    const locale = useContext(LocaleContext);
+    const { locale } = locale;
+    const { localize, options, formats } = localeMap[locale()];
+    const format = formats.javascript();
+    const formats = formats.input();
+    const [selectedDate, setSelectedDate] = useState(parseToDate(format));
+    const [open, setOpen] = useState(false));
+
+    useEffect(() => {
+      // if value string is valid format and parses to valid date setSelectedDate to parsed value
+    }, [value]);
+
+    return (
+      <div {...rest}>
+        <Textbox
+          value={value}
+          onChange={handleChange}
+          onBlur={handleBlur}
+          inputIcon="calendar"
+          iconOnClick={handleIconClick}
+          onClick={handleOnClickHandler}
+          {...inputProps}
+        />
+        <DatePicker
+          {...pickerProps}
+          value={value}
+          selectedDate={selectedDate}
+          open={open}
+          onDayClick={onDayClick}
+        />
+      </div>
+    );
+  };
+
+  DateInput.propTypes = {
+    value: PropTypes.string.isRequired,
+    onChange: PropTypes.func.isRequired,
+    onBlur: PropTypes.func.isRequired,
+    ...inputProps
+  };
```

Comment:
`...pickerProps` should be here if I've read the paragraph above right

Reply:
🤦  yep I should have added them to the prop types

Reply:
`onBlur` should not be required as well and it should be `inputPropTypes` really so I've included `pickerPropsTypes`